src/Coq.hs

```diff
@@ -0,0 +1,410 @@
+{-
+ -
+ - coq backend for act
+ -
+ - unsupported features:
+ - + bytestrings
+ - + external storage
+ - + specifications for multiple contracts
+ -
+ -}
+
+{-# Language OverloadedStrings #-}
+{-# LANGUAGE GADTs #-}
+
+module Coq where
+
+import qualified Data.Map.Strict    as M
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Text          as T
+import Data.Either (rights)
+import Data.List (find, groupBy)
+import Control.Monad.State
+
+import EVM.ABI
+import EVM.Solidity (SlotType(..))
+import Syntax
+import RefinedAst
+
+type Store = M.Map Id SlotType
+type Fresh = State Int
+
+header :: T.Text
+header =
+  "(* --- GENERATED BY ACT --- *)\n\n\
+  \Require Import Coq.ZArith.ZArith.\n\
+  \Require Import ActLib.ActLib.\n\
+  \Require Coq.Strings.String.\n\n\
+  \Module " <> strMod <> " := Coq.Strings.String.\n\
+  \Open Scope Z_scope.\n\n"
+
+-- | produce a coq representation of a specification
+coq :: Store -> [Claim] -> T.Text
+coq store claims =
+
+  header
+  <> layout store <> "\n\n"
+  <> block (evalSeq (claim store) <$> groups transitions)
+  <> block (evalSeq retVal        <$> groups transitions)
+  <> block (evalSeq (base store)  <$> groups constructors)
+  <> reachable (groups constructors) (groups transitions)
+
+  where
+
+  behaviours = [a | B a <- claims]
+
+  transitions = filter f behaviours where
+    f (Behaviour _ Pass False _ _ _ _ _ _) = True
+    f _ = False
```

Comment:
```suggestion
  transitions = filter (not . _creation) behaviours
```

Reply:
or if we introduce lenses for `Behaviour`:
`filter (not . view creation)`

Reply:
patched in https://github.com/ethereum/act/commit/48b6c628a093f6ab32fb5173ca5996e6f999fabe